client/network/src/protocol/generic_proto/handler/group.rs

```diff
@@ -302,6 +302,9 @@ impl ProtocolsHandler for NotifsHandler {
 	fn inject_event(&mut self, message: NotifsHandlerIn) {
 		match message {
 			NotifsHandlerIn::Enable => {
+				if let EnabledState::Enabled = self.enabled {
+					error!("enabling already-enabled handler");
```

Comment:
```suggestion
					debug!("enabling already-enabled handler");
```
I'm suggesting this because I really don't think it is a good idea if duplicate invocations of code which can trivially be idempotent (like enabling or disabling a handler) are treated as errors. I've been bitten by that before in substrate code in the context of https://github.com/paritytech/substrate/pull/5278 where a handler was even unnecessarily left in a poisoned state, see [here](https://github.com/paritytech/substrate/blob/master/client/network/src/protocol/generic_proto/handler/notif_out.rs#L284-L285) and [here](https://github.com/paritytech/substrate/blob/master/client/network/src/protocol/generic_proto/handler/notif_out.rs#L292-L293). I change that in #5278. These may be the causes for the other logs seen in #5201.

At worst, calling an idempotent method or callback twice is a performance issue, hence justifying debug or trace logs, but in my opinion should never be considered an error and especially not unnecessarily leave the code in a broken state.

Reply:
> At worst, calling an idempotent method or callback twice is a performance issue

My reasoning is that the logic of `src/protocol/generic_proto/behaviour.rs` is quite complicated, and a state inconsistency there can quickly degenerate into more complicated issues.

I went with `error!` in order to scream out "we've found a bug" without it going unnoticed.

Reply:
It makes sense to put avoid these poisoned state, but there is no such poisoned state in this PR, and I would leave them as `error!`s.

Reply:
> I went with error! in order to scream out "we've found a bug" without it going unnoticed.

As I said, I do not consider it a bug if an idempotent piece of code, which enabling/disabling a handler should be, is executed multiple times, but rather only a potential performance improvement if one wishes to avoid all extraneous calls. Requiring some code which could easily be idempotent to only be called "once", just makes life unnecessarily harder since it makes it harder to write functionally correct code. Again, just my two cents. The bug that is fixed here seems to me somewhere else and if multiple invocations of "enable" / "disable" would somehow be connected to a bug, `debug!` logging is sufficient to identify such unwanted interactions.

Reply:
From the perspective of the `NotificationsInHandler` it is indeed not a bug to get enabled when it is already enabled or disabled when it is already disabled.

This `error!` exists purely to detect bugs in `behaviour.rs` and it exists because of implementation details in `behaviour.rs`.

I agree that theoretically speaking this `error!` should be in `behaviour.rs` and not here, but from a pragmatic perspective, if you look at the code in `behaviour.rs`, it is much more reasonable to put it here.

Reply:
I've changed this to `debug!`s so that we hopefully get this in before the Polkadot release today.